include/glow/Support/Error.h

```diff
@@ -45,6 +45,152 @@ template <typename> struct IsLLVMExpected : public std::false_type {};
 template <typename T>
 struct IsLLVMExpected<llvm::Expected<T>> : public std::true_type {};
 
+template <typename ThisErrT> class BaseErr : public llvm::ErrorInfoBase {
+public:
+  static const void *classID() { return &ThisErrT::ID; }
+
+  const void *dynamicClassID() const override { return &ThisErrT::ID; }
+
+  bool isA(const void *const ClassID) const override {
+    return ClassID == classID() || llvm::ErrorInfoBase::isA(ClassID);
+  }
+
+  virtual std::error_code convertToErrorCode() const override {
+    return llvm::inconvertibleErrorCode();
+  }
+
+  void log(llvm::raw_ostream &OS) const override {
+    OS << "file: " << fileName_ << " line: " << lineNumber_ << " error: \"";
+    static_cast<const ThisErrT *>(this)->logErr(OS);
+    OS << "\"";
+  }
+
+protected:
+  BaseErr(llvm::StringRef fileName, int32_t lineNumber) {
+    fileName_ = fileName;
+    lineNumber_ = lineNumber;
+  }
+
+private:
+  int32_t lineNumber_;
+  std::string fileName_;
+};
+
+class ModelLoadingErr final : public BaseErr<ModelLoadingErr> {
+public:
+  // Used by BaseErr.
+  static char ID;
+
+  enum EC {
+    UNKNOWN = 0,
+    UNSUPPORTED_SHAPE = 1,
+    UNSUPPORTED_OPERATOR = 2,
+    UNSUPPORTED_ATTRIBUTE = 3,
+    UNSUPPORTED_DATATYPE = 4,
+  };
+
+  static llvm::StringRef ecToString(const EC ec) {
+    switch (ec) {
+    case UNKNOWN:
+      return "UNKNOWN";
+    case UNSUPPORTED_SHAPE:
+      return "UNSUPPORTED_SHAPE";
+    case UNSUPPORTED_OPERATOR:
+      return "UNSUPPORTED_OPERATOR";
+    case UNSUPPORTED_ATTRIBUTE:
+      return "UNSUPPORTED_ATTRIBUTE";
```

Comment:
does this work correctly?

in case if you have

const std::string& ret( ... ) {
   ...
   return "something";
}

this would return ref to the temp object.
StringRef might have similar problem

Reply:
Oh good point, I would not expect that this works actually but I could wrong. I'll change it

Reply:
If I understand StringRef correctly, this pattern should be safe, because StringRef simply wraps the (statically allocated) const char* and length, without any stack/heap allocation: https://godbolt.org/z/jQT1j8